File: Backend/vinta-academy-backend/app/utils/formatters.py
"""
Vinta School OS — Formatters & Helpers
Phone (+213), Currency (DZD), Date/Time utilities for Algeria locale.
"""
import re
from datetime import datetime, date, timedelta, timezone


# --- Phone Formatting ---

PHONE_PATTERN = re.compile(r"^(\+213|0)?(\d{9})$")


def format_phone(phone: str) -> str:
    """
    Normalize an Algerian phone number to +213XXXXXXXXX format.
    Accepts: 0555123456, 555123456, +213555123456
    Returns: +213555123456
    """
    if not phone:
        return phone

    digits = re.sub(r"[^\d]", "", phone)

    # Remove leading 0
    if digits.startswith("0") and len(digits) == 10:
        digits = digits[1:]

    # Add country code if missing
    if not digits.startswith("213"):
        digits = "213" + digits

    return f"+{digits}"


def validate_phone(phone: str) -> bool:
    """Validate an Algerian phone number format."""
    if not phone:
        return True  # Optional field
    return bool(PHONE_PATTERN.match(phone))


# --- Currency Formatting ---

def format_dzd(amount: int) -> str:
    """Format amount in Algerian Dinars with comma separators."""
    return f"{amount:,} DA"


def parse_dzd(amount_str: str) -> int:
    """Parse a DZD amount string (e.g., '3,500 DA' or '3500') to integer."""
    cleaned = re.sub(r"[^\d]", "", amount_str)
    return int(cleaned) if cleaned else 0


# --- Date/Time Utilities ---

def now_utc() -> datetime:
    """Get current UTC datetime."""
    return datetime.now(timezone.utc)


def today() -> date:
    """Get today's date in the server timezone."""
    return date.today()


def days_until(target_date: date) -> int:
    """Calculate days until a target date. Negative if past."""
    delta = target_date - today()
    return delta.days


def days_overdue(due_date: date) -> int:
    """Calculate how many days overdue a due date is. 0 if not overdue."""
    delta = today() - due_date
    return max(0, delta.days)


def overdue_bucket(days: int) -> str:
    """
    Classify overdue days into aging buckets:
    1-7d → 'recent', 8-30d → 'aging', 30d+ → 'critical'
    """
    if days <= 7:
        return "recent"
    elif days <= 30:
        return "aging"
    else:
        return "critical"


def week_date_range(reference_date=None):
    """Return (start, end) of the week (Sunday-Saturday)."""
    from datetime import date, timedelta
    t = reference_date or date.today()
    # Sunday is the start of the week
    days_since_sunday = (t.weekday() + 1) % 7
    start = t - timedelta(days=days_since_sunday)
    end = start + timedelta(days=6)
    return start, end


def month_date_range() -> tuple[date, date]:
    """Get the first and last day of the current month."""
    t = today()
    first_day = t.replace(day=1)
    if t.month == 12:
        last_day = t.replace(year=t.year + 1, month=1, day=1) - timedelta(days=1)
    else:
        last_day = t.replace(month=t.month + 1, day=1) - timedelta(days=1)
    return first_day, last_day


def next_occurrence(day_of_week: int) -> date:
    """Return the next date that falls on the given day_of_week (0=Sunday)."""
    from datetime import date, timedelta
    today = date.today()
    # Python weekday(): Monday=0 ... Sunday=6
    # Our convention: Sunday=0, Monday=1 ... Saturday=6
    # Convert: our_day 0(Sun) -> python 6, our_day 1(Mon) -> python 0, etc.
    python_day = (day_of_week - 1) % 7
    days_ahead = (python_day - today.weekday()) % 7
    if days_ahead == 0:
        days_ahead = 7  # Always return NEXT occurrence, not today
    return today + timedelta(days=days_ahead)


# --- Session Duration ---

def session_duration_hours(start_time, end_time) -> float:
    """Calculate session duration in hours from time objects."""
    from datetime import datetime as dt
    dummy = dt.today().date()
    start_dt = dt.combine(dummy, start_time)
    end_dt = dt.combine(dummy, end_time)
    delta = end_dt - start_dt
    return round(delta.total_seconds() / 3600, 2)

File: vinta-school-os/src/lib/formatters.ts
/**
 * Vinta School OS — Formatters
 * Phone, currency, date, and time formatting utilities
 */

// ============================================
// Phone Formatting
// ============================================

/**
 * Format phone number to +213 format
 * @param phone - Raw phone input
 * @returns Formatted phone with +213 prefix
 */
export function formatPhone(phone: string): string {
  // Remove all non-digit characters except +
  const cleaned = phone.replace(/[^\d+]/g, '')

  // If starts with 0, replace with +213
  if (cleaned.startsWith('0')) {
    return '+213' + cleaned.slice(1)
  }

  // If doesn't start with +213, add it
  if (!cleaned.startsWith('+213')) {
    return '+213' + cleaned
  }

  return cleaned
}

/**
 * Format phone for display: +213 5## ## ## ##
 * @param phone - Formatted phone string
 * @returns Display-formatted phone
 */
export function displayPhone(phone: string): string {
  if (!phone) return ''
  const cleaned = phone.replace(/[^\d]/g, '')
  if (cleaned.length < 10) return phone

  // +213 5XX XXX XXX
  const carrier = cleaned.slice(3, 4)
  const part1 = cleaned.slice(4, 6)
  const part2 = cleaned.slice(6, 8)
  const part3 = cleaned.slice(8, 10)

  return `+213 ${carrier}${part1} ${part2} ${part3}`
}

// ============================================
// Currency Formatting
// ============================================

/**
 * Format amount in Algerian Dinar (DA)
 * @param amount - Amount in DA
 * @returns Formatted string like "3,500 DA"
 */
export function formatCurrency(amount: number): string {
  return new Intl.NumberFormat('fr-DZ', {
    style: 'decimal',
    minimumFractionDigits: 0,
    maximumFractionDigits: 0,
  }).format(amount) + ' DA'
}

/**
 * Integer-only DZD helper — currency locked to DZD display.
 * Accepts floats but rounds to the nearest integer (no centimes).
 * @param amount - Amount in DZD (integer expected)
 * @returns Formatted string like "2000 Da"
 */
export function formatDa(amount: number): string {
  const int = Math.round(Number.isFinite(amount) ? amount : 0)
  return `${new Intl.NumberFormat('fr-DZ', {
    style: 'decimal',
    minimumFractionDigits: 0,
    maximumFractionDigits: 0,
  }).format(int)} Da`
}

/**
 * Format amount without currency symbol
 * @param amount - Amount
 * @returns Formatted string like "3,500"
 */
export function formatAmount(amount: number): string {
  return new Intl.NumberFormat('fr-DZ', {
    style: 'decimal',
    minimumFractionDigits: 0,
    maximumFractionDigits: 0,
  }).format(amount)
}

// ============================================
// Date Formatting
// ============================================

/**
 * Format date as YYYY-MM-DD
 * @param date - Date object or string
 * @returns Formatted date string
 */
export function formatDateISO(date: Date | string): string {
  const d = typeof date === 'string' ? new Date(date) : date
  return d.toISOString().split('T')[0]
}

/**
 * Format date as "Monday, January 5, 2026"
 * @param date - Date object or string
 * @returns Full date string
 */
export function formatDateFull(date: Date | string): string {
  const d = typeof date === 'string' ? new Date(date) : date
  return d.toLocaleDateString('en-US', {
    weekday: 'long',
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  })
}

/**
 * Format date as "Jan 5" or "5 Jan"
 * @param date - Date object or string
 * @param short - If true, use short month
 * @returns Short date string
 */
export function formatDateShort(date: Date | string, short = true): string {
  const d = typeof date === 'string' ? new Date(date) : date
  return d.toLocaleDateString('en-US', {
    month: short ? 'short' : 'long',
    day: 'numeric',
  })
}

/**
 * Get day name (Sun, Mon, Tue, etc.)
 * @param date - Date object or string
 * @returns Day name
 */
export function getDayName(date: Date | string, short = true): string {
  const d = typeof date === 'string' ? new Date(date) : date
  return d.toLocaleDateString('en-US', {
    weekday: short ? 'short' : 'long',
  })
}

/**
 * Get week dates (Sun-Sat)
 * @param startDate - Any date in the week
 * @returns Array of 7 Date objects
 */
export function getWeekDates(startDate: Date): Date[] {
  const dates: Date[] = []
  const start = new Date(startDate)
  // Get to Sunday (day 0)
  start.setDate(start.getDate() - start.getDay())

  for (let i = 0; i < 7; i++) {
    dates.push(new Date(start))
    start.setDate(start.getDate() + 1)
  }

  return dates
}

/**
 * Check if two dates are the same day
 */
export function isSameDay(a: Date | string, b: Date | string): boolean {
  const dateA = typeof a === 'string' ? new Date(a) : a
  const dateB = typeof b === 'string' ? new Date(b) : b
  return (
    dateA.getFullYear() === dateB.getFullYear() &&
    dateA.getMonth() === dateB.getMonth() &&
    dateA.getDate() === dateB.getDate()
  )
}

/**
 * Check if date is today
 */
export function isToday(date: Date | string): boolean {
  return isSameDay(date, new Date())
}

// ============================================
// Time Formatting
// ============================================

/**
 * Format time string "HH:MM" to decimal hours
 * @param time - Time string like "08:30"
 * @returns Decimal hours like 8.5
 */
export function timeToDecimal(time: string): number {
  const [hours, minutes] = time.split(':').map(Number)
  return hours + minutes / 60
}

/**
 * Format decimal hours to "12 PM" style
 * @param hours - Decimal hours
 * @returns Formatted time like "12 PM"
 */
export function formatHour12(hours: number): string {
  const h = Math.floor(hours)
  const ampm = h >= 12 ? 'PM' : 'AM'
  const h12 = h > 12 ? h - 12 : h === 0 ? 12 : h
  return `${h12} ${ampm}`
}

/**
 * Format decimal hours to "1:30PM" style
 * @param hours - Decimal hours
 * @returns Formatted time like "1:30PM"
 */
export function formatTime12(hours: number): string {
  const h = Math.floor(hours)
  const m = Math.round((hours - h) * 60)
  const ampm = h >= 12 ? 'PM' : 'AM'
  const h12 = h > 12 ? h - 12 : h === 0 ? 12 : h
  return m > 0 ? `${h12}:${m.toString().padStart(2, '0')}${ampm}` : `${h12}${ampm}`
}

/**
 * Format time string "HH:MM" to "1:30 PM"
 * @param time - 24h time string
 * @returns 12h formatted time
 */
export function formatTime(time: string): string {
  const decimal = timeToDecimal(time)
  return formatTime12(decimal)
}

/**
 * Get current time as decimal hours
 */
export function getCurrentHour(): number {
  const now = new Date()
  return now.getHours() + now.getMinutes() / 60
}

// ============================================
// Initials
// ============================================

/**
 * Extract initials from name (max 2 chars)
 * @param name - Full name
 * @returns Initials like "AB"
 */
export function getInitials(name: string): string {
  if (!name) return ''
  const parts = name.trim().split(/\s+/)
  if (parts.length === 1) {
    return parts[0].charAt(0).toUpperCase()
  }
  return (parts[0].charAt(0) + parts[parts.length - 1].charAt(0)).toUpperCase()
}

// ============================================
// Days Overdue
// ============================================

/**
 * Calculate days overdue from due date
 * @param dueDate - Due date string
 * @returns Number of days overdue (0 if not overdue)
 */
export function getDaysOverdue(dueDate: string): number {
  const due = new Date(dueDate)
  const today = new Date()
  today.setHours(0, 0, 0, 0)
  due.setHours(0, 0, 0, 0)

  const diff = today.getTime() - due.getTime()
  return Math.max(0, Math.ceil(diff / (1000 * 60 * 60 * 24)))
}

/**
 * Get aging bucket for overdue days
 * @param daysOverdue - Number of days overdue
 * @returns 'recent' | 'aging' | 'critical'
 */
export function getAgingBucket(daysOverdue: number): 'recent' | 'aging' | 'critical' {
  if (daysOverdue <= 7) return 'recent'
  if (daysOverdue <= 30) return 'aging'
  return 'critical'
}

// ============================================
// Status Helpers
// ============================================

/**
 * Get status color class
 * @param status - Payment/attendance status
 * @returns Tailwind color class
 */
export function getStatusColor(status: string): string {
  switch (status) {
    case 'paid':
    case 'present':
    case 'completed':
      return 'text-emerald'
    case 'due':
    case 'scheduled':
      return 'text-gold'
    case 'overdue':
    case 'absent':
    case 'cancelled':
      return 'text-red'
    default:
      return 'text-muted'
  }
}

/**
 * Get status background class
 * @param status - Payment/attendance status
 * @returns Tailwind background class
 */
export function getStatusBg(status: string): string {
  switch (status) {
    case 'paid':
    case 'present':
    case 'completed':
      return 'bg-emerald-soft'
    case 'due':
    case 'scheduled':
      return 'bg-gold-soft'
    case 'overdue':
    case 'absent':
    case 'cancelled':
      return 'bg-red-soft'
    default:
      return 'bg-gray-100'
  }
}

// ============================================
// Number Formatting
// ============================================

/**
 * Format number with locale-specific separators
 * @param num - Number to format
 * @returns Formatted number string
 */
export function formatNumber(num: number): string {
  return new Intl.NumberFormat('en-US').format(num)
}

/**
 * Format percentage
 * @param value - Decimal value (0.5 = 50%)
 * @returns Formatted percentage string
 */
export function formatPercent(value: number): string {
  return `${Math.round(value * 100)}%`
}
